lightning/src/routing/network_graph.rs

```diff
@@ -428,6 +428,71 @@ impl Readable for NodeInfo {
 	}
 }
 
+/// Allows for updating user's network metadata.
+pub trait RouteFeePenalty {
+	/// Gets a channel's fee penalty based on its channel_id (as stored in a NetworkGraph object).
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> u64;
+	/// Informs metadata object that a route has successfully executed its payment.
+	fn route_succeeded(&mut self, route: Vec<RouteHop>);
+	/// Informs metadata object that a route has failed to execute a payment.
+	fn route_failed(&mut self, route: Vec<RouteHop>, failed_hop: RouteHop);
+}
+
+/// A default metadata object that is used to implement the default functionality for the
+/// NetworkTracker trait. A user could  make their own Metadata object and extend the
+/// functionality of it by implementing other functions/traits for their metadata.
+pub struct DefaultMetadata {
+	/// A lits of failed channels. Maps channel_id (as specified in the NetworkGraph object) to
+	/// the number of successful routes to participate before being removed from the list. All
+	/// channels in failed_channels are assumed to have a penalty of u64::max.
+	failed_channels: BTreeMap<u64, u64>,
+}
+
+impl RouteFeePenalty for DefaultMetadata {
+	fn get_channel_fee_penalty(&self, chan_id: u64) -> u64 {
+		if self.failed_channels.get(&chan_id) == None {
+			return 0;
+		} else {
+			return u64::max_value();
+		}
+	}
+
+	fn route_succeeded(&mut self, route: Vec<RouteHop>) {
+		for route_hop in route {
+			let chan_id = route_hop.short_channel_id;
+			let successes_needed = self.failed_channels.get(&chan_id);
+			if successes_needed == None {
+				self.failed_channels.insert(chan_id, 5);
```

Comment:
I'm super confused here. Looks like you're inserting channels which will be considered "failed" in get_channel_fee_penalty on the basis that they succeeded?

Reply:
Yes, that's right. It's definitely a bug, I removed this branch of the if statement